File: 자료구조/나영/BOJ_9012_괄호.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static Stack<Character> stack = new Stack<>();
    static int T;
    static char[] charn;
    static boolean isS;
    static String result;
    public static void main(String[] args) throws Exception {
        T = Integer.parseInt(br.readLine());

        for (int t = 0; t < T; t++) {
            charn = br.readLine().toCharArray();
            stack = new Stack<>();
            isS = true;
            for (int i = 0; i < charn.length; i++) {
                if (charn[i] == '(') stack.add(charn[i]);
                else if (!stack.isEmpty()) stack.pop();
                else isS = false;
            }
            
            if (!isS || stack.size()!=0) result = "NO";
            else result = "YES";            
            System.out.println(result);
        }
    }
}

File: 자료구조/지우/BOJ_9012_괄호.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {
    int T;
    cin >> T;

    for(int tc=1; tc<=T; tc++) {
        string answer = "YES";
        
        string s;
        cin >> s; 

        stack<int> stack;
        for(int i=0; i<s.length(); i++) {
            if(s[i] == '(') {
                stack.push(1);
            } else {
                if(stack.empty()) {
                    // 게임 끝
                    answer = "NO";
                    break;
                } else {
                    stack.pop();
                }
            }
        }

        if(stack.empty() && answer == "YES") {
            cout << answer << "\n";
        } else {
            cout << "NO" << "\n";
        }
    }
}